File: libcirc/noapprox.py
#
# file: noapprox.py
# Calaculate || \Pi |H^(\otimes t)> exactly
#

import numpy as np
from libcirc.stabilizer.stabilizer import StabilizerState


def prepH(i, t):
    size = int(np.ceil(t/2))
    odd = t % 2 == 1

    bits = list(np.binary_repr(i, width=size))

    # initialize stabilizer state
    phi = StabilizerState(t, t)

    # set J matrix
    for idx in range(size):
        bit = int(bits[idx])
        if bit == 0 and not (odd and idx == size-1):
            # phi.J = np.array([[0, 4], [4, 0]])
            phi.J[idx*2+1, idx*2] = 4
            phi.J[idx*2, idx*2+1] = 4

    # truncate affine space using shrink
    for idx in range(size):
        bit = int(bits[idx])
        vec = np.zeros(t)

        if odd and idx == size-1:
            # bit = 0 is |+>
            # bit = 1 is |0>
            if bit == 1:
                vec[t-1] = 1
                phi.shrink(vec, 0)
            continue

        # bit = 1 corresponds to |00> + |11> state
        # bit = 0 corresponds to |00> + |01> + |10> - |11>

        if bit == 1:
            vec[idx*2] = 1
            vec[idx*2+1] = 1
            phi.shrink(vec, 0)  # only 00 and 11 have inner prod 0 with 11

    return phi


def calcExactProjector(P, Lnorm):
    (phases, xs, zs) = P

    # empty projector
    if len(phases) == 0:
        return np.abs(Lnorm)**2

    t = len(xs[0])

    # clifford circuit
    if t == 0:
        lookup = {0: 1, 2: -1}
        generators = [1]  # include identity
        for phase in phases: generators.append(lookup[phase])

        # calculate sum of all permutations of generators
        return sum(generators)/len(generators)

    size = int(np.ceil(t/2))
    total = 0
    for i in range(0, 2**size):
        theta = prepH(i, t)
        projfactor = 1
        for g in range(len(phases)):
            res = theta.measurePauli(phases[g], zs[g], xs[g])
            projfactor *= res

            if res == 0: break  # theta annihilated by P
        if projfactor == 0: continue

        for j in range(0, 2**size):
            phi = prepH(j, t)
            inner = StabilizerState.innerProduct(theta, phi)
            total += inner * projfactor

    return np.abs(total)

File: libcirc/noapprox.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
// #include "stabilizer/stabilizer.h"
#include "mpi.h"
#include "utils/utils.h"

struct StabilizerState* prepH(int i, int t) {
	int size = ceil((double)t/2);
	
    char buff[size+1];
	char *bits = binrep(i,buff,size);
	
	struct StabilizerState *phi = allocStabilizerState(t, t);

    // set J matrix
	for(int j=0;j<size;j++){
		if(bits[j] == '0' && !(t%2 && j==size-1)){
			gsl_matrix_set(phi->J, j*2+1, j*2, 4);
			gsl_matrix_set(phi->J, j*2, j*2+1, 4);
		}
	}

	gsl_vector *tempVector;
	tempVector = gsl_vector_alloc(t);
	
	for(int j=0;j<size;j++){
		gsl_vector_set_zero(tempVector);
		
		if(t%2 && j==size-1){
			gsl_vector_set(tempVector, t-1, 1);
		
            // bit = 0 is |+>
            // bit = 1 is |0>
			if(bits[j] == '1'){
			    shrink(phi, tempVector, 0, 0);	//|0>
			}
			
			continue;
		}
        
	    // bit = 1 corresponds to |00> + |11> state
        // bit = 0 corresponds to |00> + |01> + |10> - |11>
		if(bits[j] == '1'){
            gsl_vector_set(tempVector, j*2+1, 1);
            gsl_vector_set(tempVector, j*2, 1);
        
            shrink(phi, tempVector, 0, 0); // only 00 and 11 have inner prod 0 with 11
		}
	}

    return phi;
}

double exactProjector(struct Projector *P, double Lnorm) {
    printf("Lnorm: %f\n", Lnorm);
    printf("Projector\n");
    printProjector(P);
    // empty projector
    if (P->Nstabs == 0) return Lnorm;

    int t = P->Nqubits;

    // clifford circuit
    if (t == 0) {
        double sum = 1; // include identity
        for (int i = 0; i < P->Nstabs; i++) {
            double ph = gsl_vector_get(P->phases, i);
            if (ph == 0) sum += 1;
            if (ph == 2) sum -= 1;
        } 

        return sum/(1 + (double)P->Nstabs);
    }

    // exact inner product
    int size = pow(2, ceil((double)t / 2));
    gsl_complex total = gsl_complex_rect(0,0);
    for (int i = 0; i < size; i++) {
        struct StabilizerState* theta = prepH(i, t);

        // Project theta
        double projfactor = 1;
        gsl_vector *zeta = gsl_vector_alloc(P->Nqubits);
        gsl_vector *xi = gsl_vector_alloc(P->Nqubits);

        for (int j = 0; j < P->Nstabs; j++) {
            int m = gsl_vector_get(P->phases, i);
            gsl_matrix_get_row(zeta, P->zs, i);
            gsl_matrix_get_row(xi, P->xs, i);

            double res = measurePauli(theta, m, zeta, xi);
            projfactor *= res;

            if (res == 0) break;
        } 
        
        gsl_vector_free(zeta);
        gsl_vector_free(xi);

        if (projfactor == 0) continue;
        
        // Evaluate other components
        for (int j = 0; j < size; j++) {
            struct StabilizerState* phi = prepH(j, t);
            gsl_complex innerProd;
            
            int eps, p, m;
        	innerProduct(theta, phi, &eps, &p, &m, &innerProd, 0);
            freeStabilizerState(phi);
            total = gsl_complex_add(total, innerProd);  
        }

        total = gsl_complex_mul_real(total, projfactor);
    }

    return gsl_complex_abs(total);
}
